### PR TITLE
Implementation of Relative indexing method proposal

### PR DESCRIPTION
**Pull request by MadProbe**
Relative indexing method proposal is stage 3
https://github.com/tc39/proposal-relative-indexing-method

Implementation includes:
- [x] Array.prototype.at
- [x] %TypedArray%.prototype.at
- [x] String.prototype.at
- [x] Array.prototype[@@unscopables].at entry
- [x] Tests

### CODE REVIEW

**Comment by rhuanjl** [comment]
This looks great - it will take me a bit of time to review, and need to solve why it's failing with undefined/missing symbols on 3 of the builds.

File: lib/Runtime/Library/JavascriptArray.cpp [hunk 1]
```diff
@@ -8770,6 +8771,89 @@ using namespace Js;
...
+
+        if (length.IsSmallIndex())
+        {
+            JS_REENTRANT_UNLOCK(jsReentLock, return JavascriptArray::AtHelper(pArr, nullptr, obj, length.GetSmallIndex(), args, scriptContext));
+        }
+        Assert(pArr == nullptr || length.IsUint32Max()); // if pArr is not null lets make sure length is safe to cast, which will only happen if length is a uint32max
+        JS_REENTRANT_UNLOCK(jsReentLock, return JavascriptArray::AtHelper(pArr, nullptr, obj, length.GetBigIndex(), args, scriptContext));
+    }
+
+    template <typename T>
```
Comment (rhuanjl) [inline lib/Runtime/Library/JavascriptArray.cpp:8812]: Why does this need to be a template method? (I think the CI failures are to do with failing to instantiate some forms of this template)

File: lib/Runtime/Library/JavascriptArray.cpp [hunk 2]
```diff
@@ -8770,6 +8771,89 @@ using namespace Js;
...
+        {
+            return scriptContext->GetLibrary()->GetUndefined();
+        }
+        
+        
+        if (typedArrayBase)
+        {
+            // %typedarray%.prototype.at(index): https://tc39.es/proposal-relative-indexing-method/#sec-array.prototype.at
+            // 8. Return ? Get(O, ! ToString(k)).
+            return typedArrayBase->DirectGetItem((uint32_t)k);
```
Comment (rhuanjl) [inline lib/Runtime/Library/JavascriptArray.cpp:8848]: What happens if `TypedArray.prototype` has a getter on it or has been turned into a proxy?

File: lib/Runtime/Library/JavascriptArray.cpp [hunk 3]
```diff
@@ -8770,6 +8771,89 @@ using namespace Js;
...
+        if (args.Info.Count == 0)
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_This_NullOrUndefined, _u("Array.prototype.at"));
+        }
+
+        BigIndex length;
+        JavascriptArray* pArr = nullptr;
+        RecyclableObject* obj = nullptr;
+
+        JS_REENTRANT(jsReentLock, TryGetArrayAndLength(args[0], scriptContext, _u("Array.prototype.at"), &pArr, &obj, &length));
```
Comment (rhuanjl) [inline lib/Runtime/Library/JavascriptArray.cpp:8801]: What happens if someone does `Array.prototype.at.call(not_an_array, some_int);`?

File: lib/Runtime/Library/JavascriptString.cpp [hunk 4]
```diff
@@ -696,6 +697,70 @@ namespace Js
...
+        {
+#ifdef ENABLE_SPECTRE_RUNTIME_MITIGATIONS
+            value = BreakSpeculation(value);
+#endif
+            return value;
+        }
+        else
+        {
+            // Edge case?
+            return scriptContext->GetLibrary()->GetEmptyString();
```
Comment (rhuanjl) [inline lib/Runtime/Library/JavascriptString.cpp]: Do you expect to hit this case? What test case does it?

File: lib/Runtime/LibraryFunction.h [hunk 5]
```diff
@@ -19,6 +20,7 @@ LIBRARY_FUNCTION(Math,          Asin,               1,    BIF_TypeSpecUnaryToFlo
 LIBRARY_FUNCTION(Math,          Atan,               1,    BIF_TypeSpecUnaryToFloat                          , Math::EntryInfo::Atan)
 LIBRARY_FUNCTION(Math,          Atan2,              2,    BIF_TypeSpecAllToFloat                            , Math::EntryInfo::Atan2)
 LIBRARY_FUNCTION(Math,          Ceil,               1,    BIF_TypeSpecDstToInt | BIF_TypeSpecSrc1ToFloat    , Math::EntryInfo::Ceil)
+LIBRARY_FUNCTION(JavascriptString,        At,                 2,    BIF_UseSrc0                                       , JavascriptString::EntryInfo::At)
```
Comment (rhuanjl) [inline lib/Runtime/LibraryFunction.h:23]: This file specifically lists functions that can be inlined by the JIT - if we're going to add these here we should also add Jit handlers for them in Backend/InliningDecider.cpp

It would be nice to give a jit fastpath if we can to make these methods match (or nearly match) the speed of normal property access.

**Comment by rhuanjl** [comment]
The test cases don't seem to look at what happens when you call these methods with inappropriate targets either with .call/.apply OR by copying the methods OR by overwriting/editing TypdArray.prototype or Array.prototype or String.prototype - I don't think the current implementations are likely to handle all of those cases correctly.

(Awkwardly with methods like this, complying with the spec when someone uses a method incorrectly is often the trickiest bit)

Are there any Test262 test cases we can check against as well?

File: lib/Runtime/Library/JavascriptArray.cpp (hunk 3 above)
Comment (MadProbe) [reply lib/Runtime/Library/JavascriptArray.cpp:8801]: Method will get the object and the length property of it, and then all is on algorithm.
This is valid:
```js
[].at.call({ 0: 1, 1: 2, length: 2 }, 0); // 1
```

File: lib/Runtime/Library/JavascriptArray.cpp (hunk 1 above)
Comment (MadProbe) [reply lib/Runtime/Library/JavascriptArray.cpp:8812]: I think, you guessed wrong - then EntryHelper also shouldn't to "instantiate some forms of this template".
https://github.com/chakra-core/ChakraCore/blob/master/lib/Runtime/Library/JavascriptArray.cpp#L8773-L8883
(Sorry, if this statment sounds sarcastic.)

File: lib/Runtime/Library/JavascriptArray.cpp (hunk 2 above)
Comment (MadProbe) [reply lib/Runtime/Library/JavascriptArray.cpp:8848]: Probably, just as written here (code).
```js
{
    const a = new Int16Array;
    Object.defineProperty(Object.getPrototypeOf(Int16Array).prototype, "length", {
        get: () => 100
    });
    Object.defineProperty(Object.getPrototypeOf(Int16Array).prototype, "0", {
        get: () => 100
    });
    print(a.at(0)); // Still undefined
    print(new Int16Array([123]).at(0)) // Still 123
}
// Same in Chrome Canary
```

Comment (MadProbe) [reply lib/Runtime/Library/JavascriptArray.cpp:8848]: (Maybe it's not right to look at v8's impl, because it also can be wrong)

File: lib/Runtime/Library/JavascriptString.cpp (hunk 4 above)
Comment (MadProbe) [reply lib/Runtime/Library/JavascriptString.cpp]: It doesn't, just left this

File: lib/Runtime/LibraryFunction.h (hunk 5 above)
Comment (MadProbe) [reply lib/Runtime/LibraryFunction.h:23]: ok, i'll add them here when i will understand where i should add them

File: lib/Runtime/Library/JavascriptArray.cpp (hunk 1 above)
Comment (rhuanjl) [reply lib/Runtime/Library/JavascriptArray.cpp:8812]: The macOS X build failed with:
Undefined symbols for architecture x86_64:
  "void* Js::JavascriptArray::AtHelper<unsigned int>(Js::JavascriptArray*, Js::TypedArrayBase*, Js::RecyclableObject*, unsigned int, Js::Arguments&, Js::ScriptContext*)", referenced from:
      Js::TypedArrayBase::EntryAt(Js::RecyclableObject*, Js::CallInfo, ...) in libChakraCoreStatic.a(TypedArray.cpp.o)

Perhaps need to make the template instantiation explicit.

**Comment by MadProbe** [comment]
There are.

File: lib/Runtime/Library/JavascriptArray.cpp (hunk 2 above)
Comment (rhuanjl) [reply lib/Runtime/Library/JavascriptArray.cpp:8848]: The spec text uses `Get` which ought to hit getters - though it's targeting the ViewedArrayBuffer internal slot rather than the typed array itself - not actually 100% what the correct logic is in that case - would be good to see if there's a Test262 case on it (and suggest that there should be if there isn't) then implement whatever will pass that test.

File: lib/Runtime/Library/JavascriptArray.cpp (hunk 3 above)
Comment (rhuanjl) [reply lib/Runtime/Library/JavascriptArray.cpp:8801]: OK - should include some test cases that provide sensible objects (like that) AND some with objects that are incompatible too to test that it fails the right way.

File: lib/Runtime/Library/JavascriptString.cpp (hunk 4 above)
Comment (rhuanjl) [reply lib/Runtime/Library/JavascriptString.cpp]: If it's impossible swap the `if` to an assertion and get rid of the else?

Comment (MadProbe) [reply lib/Runtime/Library/JavascriptString.cpp]: ok,ok

File: lib/Runtime/Library/JavascriptArray.cpp (hunk 1 above)
Comment (MadProbe) [reply lib/Runtime/Library/JavascriptArray.cpp:8812]: Explicit template instantiation means something like that?
```cpp
template 
 Var JavascriptArray::AtHelper<uint32_t>(JavascriptArray* pArr, Js::TypedArrayBase* typedArrayBase, RecyclableObject* obj, uint32_t length, Arguments& args, ScriptContext* scriptContext);
```

File: lib/Runtime/Library/JavascriptString.cpp [hunk 6]
```diff
@@ -696,6 +697,63 @@ namespace Js
...
+        }
+        
+        // 7. If k < 0 or k >= len, then return undefined.
+        if (k < 0 || k >= (int64_t)len)
+        {
+            return scriptContext->GetLibrary()->GetUndefined();
+        }
+        
+        Var value;
+        Assert(pThis->GetItemAt((charcount_t)k, &value));
```
Comment (rhuanjl) [inline lib/Runtime/Library/JavascriptString.cpp]: Asserts are compiled out on release builds so need to do the Get on the line before the Assert - this is the reason for the new build failures.

Comment (MadProbe) [reply lib/Runtime/Library/JavascriptString.cpp]: Is change from Assert to AssertOrFailFast ok?
I think yes, because other case shouldn't happen at all

Comment (rhuanjl) [reply lib/Runtime/Library/JavascriptString.cpp:750]: The test failure just before you pushed the latest update appeared to be hitting the FailFast you'd added - be good to work out what case could trigger this condition if it's not meant to happen.

File: lib/Runtime/Library/JavascriptArray.cpp (hunk 1 above)
Comment (rhuanjl) [reply lib/Runtime/Library/JavascriptArray.cpp:8812]: With the update you've done it builds on macOS now - though TBH I just get confused with template instantiation sometimes seeing as the existing SliceHelper was ok and set up the same way also - your original code would build with gcc but not with clang - I guess clang is just a bit too finicky sometimes.

File: lib/Runtime/Library/JavascriptString.cpp (hunk 6 above)
Comment (rhuanjl) [reply lib/Runtime/Library/JavascriptString.cpp]: So MSVC doesn't like the possibility of the Var value being undefined.

Looking at it GetItemAt should only leave you undefined if k is > length - which you've already checked above.

So I'd just put this:
```c
Var value = nullptr;
pThis->GetItemAt((charcount_t)k, &value);
Assert(value != nullptr);
```

Comment (MadProbe) [reply lib/Runtime/Library/JavascriptString.cpp]: i'll put that if revert of this if will fail.

**Comment by MadProbe** [comment]
What's the difference between intenpreted tests and dynapogo (what does that word mean?) ones?

File: lib/Backend/InliningDecider.cpp [hunk 7]
```diff
@@ -505,12 +506,14 @@ bool InliningDecider::GetBuiltInInfoCommon(
...
         break;
 
+    case Js::JavascriptBuiltInFunction::JavascriptArray_At:
     case Js::JavascriptBuiltInFunction::JavascriptArray_Concat:
     case Js::JavascriptBuiltInFunction::JavascriptArray_Reverse:
     case Js::JavascriptBuiltInFunction::JavascriptArray_Shift:
     case Js::JavascriptBuiltInFunction::JavascriptArray_Slice:
     case Js::JavascriptBuiltInFunction::JavascriptArray_Splice:
 
+    case Js::JavascriptBuiltInFunction::JavascriptString_At:
```
Comment (rhuanjl) [inline lib/Backend/InliningDecider.cpp:516]: To make this work need to track it through further - would go from here through to inline.cpp also would need helpers in JnHelperMethodList.h.

Alternatively we can do without inliner support - delete these entries AND delete the entries in LibraryFunction.h

**Comment by rhuanjl** [comment]
> What's the difference between interpreted tests and dynapogo (what does that word mean?) ones?

Dynopogo tries to run the FullJit and trigger every possible optimisation - in this case the crash is probably hitting the inliner as you've done the first 2 steps of inlining support but there are a few more steps to make it work.

File: lib/Backend/InliningDecider.cpp (hunk 7 above)
Comment (MadProbe) [reply lib/Backend/InliningDecider.cpp:516]: I better work some more but get the optimization

**Comment by rhuanjl** [comment]
This looks good overall - thank you very much for the contribution - I'm going to run it against test262 later and check a few other edge cases before merging.

I may also look at a faster fast path - currently this will be good in most cases BUT if used in a performance critical loop could be measurably slower than normal property access - would be nice to equal it if we can.

**Review by rhuanjl — APPROVED** [review]

**Comment by rhuanjl** [comment]
Passed the test262 tests and edge cases I was interested in behaviour is the same as v8 and jsc so happy to go with it.

I note that we didn't put a feature flag on this - I'm 50/50 on the need for feature flags going forward now this isn't used in a browser, can revisit later if needed.
